src/firewall/core/fw_policy.py

```diff
@@ -39,20 +39,44 @@ def new_transaction(self):
 
     # policies
 
-    def get_policies_not_derived_from_zone(self):
-        return sorted(
-            (p.name for p in self._policies.values() if not p.derived_from_zone)
-        )
-
-    def get_active_policies_not_derived_from_zone(self):
-        active_policies = []
-        for policy in self.get_policies_not_derived_from_zone():
-            p_obj = self.get_policy(policy)
-            if (set(p_obj.ingress_zones) & set(self._fw.zone.get_active_zones() + ["HOST", "ANY"])) and \
-               (set(p_obj.egress_zones)  & set(self._fw.zone.get_active_zones() + ["HOST", "ANY"])):
-                active_policies.append(policy)
-
-        return active_policies
+    def _policy_get_active_zones_set(self):
+        return set(self._fw.zone.get_active_zones() + ["HOST", "ANY"])
+
+    def policy_is_active(self, policy, active_zones=None):
+        if policy is None:
+            return False
+        if policy.derived_from_zone:
+            return False
+        if active_zones is None:
+            active_zones = self._policy_get_active_zones_set()
+        if active_zones.isdisjoint(policy.ingress_zones):
+            return False
+        if active_zones.isdisjoint(policy.egress_zones):
+            return False
+        return True
+
+    def get_policies(
+        self,
+        *,
+        require_not_derived_from_zone=True,
+        require_active=False,
+        sort=True,
```

Comment:
I don't like this merge of functions. I strongly prefer multiple functions over a single one with multiple keyword arguments that change the behavior. It violates the "do one thing and do it well" principle.

Reply:
> So this improves one violation just to introduce another.

There is no violation. Those objects all have a `self._fw` attribute and they use it heavily. Basically, always when they need to access something globally. Having one more place (of 200+) that make use of `self._fw` at a place where it's needed, is fine.

Reply:
> I don't like this merge of functions. I strongly prefer multiple functions over a single one with multiple keyword arguments that change the behavior. It violates the "do one thing and do it well" principle.

There are 3 optional arguments.

- `sort`, which is a convenience to simplify
```python
     policies = list(self.policy.get_policies())
     policies.sort()
```
Note that the sorting provides a defined order, and in practice most callers want that.

- two filter arguments `require_not_derived_from_zone`, `require_active`. Note that the meaning of both arguments is "limiting/filtering". Meaning, setting them to TRUE strictly filters out elements. That makes it understandable, and also makes the two arguments composable in a way that makes sense. Since sort is a useful functionality, we want to filter *before* sort.

Also, the idea that more functions (e.g. `ipaddr_get_netmask_v4(plen)`, `ipaddr_get_netmask_v6(plen)`) is better than having function configurable via arguments (`ipaddr_get_netmask(addr_family, plen)` is often not true. Arguments are great, especially on python, where we can have sensible default values and have keyword-only arguments (requiring the caller to spell it out).